s30020-NikiforKoseda/zad2.txt

```diff
@@ -0,0 +1,226 @@
+RANDOM
+### Generuje losową liczbę - genialna sprawa. Nie sposób zliczyć do czego może się przydać, chociażby do losowania tapety.
+PWD
+### Przechowuje ścieżkę folderu w którym się znajdujemy. Myślę że może się przydać w skryptach tworzących coś w bierzącym folderze
+PS1
+### Przechowuje wygląd stringa na lewo od naszego obszaru wpisywania - czyli definiuje to co poprzedza to co piszemy.
+USER
+### Przechowuje nazwę aktualnego użytkownika. Przydatne by sprawdzić, czy np. użytkownik wywołujący polecenie pokrywa się z aktualnym użytkownikiem
+HOSTNAME
+### Przechowuje nazwę hosta. Nie mam do końca pomysłu do czego może być to przydatne - może żeby sprawdzić czy podpięliśmy się do odpowiedniego serwera?
+OSTYPE
+### Przechowuje typ systemu operacyjnego. Przydatne by poznać na czym pracujemy
+
+4. Wykonaj komendę ls -l w podpowłoce i przypisz ją do zmiennej X.
+
+Wyświetl zawartość tej zmiennej.
+
+Spraw, aby nie było różnicy (co najwyżej kolorki) między zwykłym wykonaniem komendy:
+
+ls -l
+
+a wypisaniem zawartości zmiennej X. Nie będzie dobrego rezultatu gdy zrobimy tak:
+
+echo $X
+
+Zastanów się dlaczego.
+
+### Wystarczy napisać echo "$X". Ze zdobytej przez mnie wiedzy wynika, że bez cudzysłowia bash traktuje $X jako listę oddzieloną spacjami, więc wszystkie whitespace'y zostaną skasowane i printuje się to jako jeden mooonstrualny teskt. Natomiast użycie "" sprawia, że wszystko zostaje zachowane. Może to przez to, że cudzysłów nakazuje interpretację wszystkich znaków i zmiennych,więc znak nowej lini również jest interpretowany?
+
+Ćwiczenie 5a
+Pamiętasz jedno z zadań poprzednich?
+
+A=Ala
+
+echo $A ma kota, a kot ma ${A}ę
+
+Przerób je tak, aby tekst wypisywał się poprawnie.
+
+Napisz skrypt, który będzie wypisywał taki tekst także dla imion męskich. Pomijamy zdrobnienia i imiona które się dziwnie odmieniają (na przykład Marek). Niech skrypt ten przyjmuje jeden argument - imię. Przykładowe wywołanie:
+
+$ ./pszetżkole Ala
+
+Ala ma kota, a kot ma Alę
+
+$ ./pszetżkole Justyna
+
+Justyna ma kota, a kot ma Justynę
+
+$ ./pszetżkole Stefan
+
+Stefan ma kota, a kot ma Stefana
+
+Czy jest to możliwe w czystym bash-u?
+
+###
+W czystym bashu na pewno możliwe jest to:
+
+#!/bin/bash
+imie=$1
+nowe_imie="${imie:0:-1}ę"
+echo "$imie ma kota, a kot ma $nowe_imie"
+
+Jeżeli chodzi o męskie to:
+
+#!/bin/bash
+imie=$1
+ostatnia_litera="${imie: -1}"
+
+if [ "$ostatnia_litera" == "a" ]; then
+  nowe_imie="${imie:0:-1}ę"
+  echo "$imie ma kota, a kot ma $nowe_imie"
+else
+  nowe_imie="${imie}a"
+  echo "$imie ma kota, a kot ma $nowe_imie"
+fi
+
+Więc tak, da się to zrobić i działa. Ciężej z tymi inaczej odmienianymi imionami.
+Podczas tworzenia natrafiłem na błąd, którego nie potrafię wyjaśnić. Jeżeli echo znajdowało się poza ifem to drugie męskie imie się nie drukowało. Nie byłem w stanie zidentyfikować problemu. Czy zmienne w ifach są dostępne poza nimi?
```

Comment:
To tak jak tworzenie zmiennych wewnątrz bloku/instrukcji_blokowej/klamr w ANSI C89. Poza blokiem dana definicja nie jest już dostępna. Więc jeśli przed if-em wstawisz sobie definicję zmiennej i zainicjujesz ją na jakąś wartość to wewnątrz bloku i poniżej niego będziesz mógł z niej skorzystać. Jest to także uznawane za dobrą praktykę. Od następnych zajęć zaczynamy na ćwiczeniach programowanie w ANSI C89 i tam to będzie wymóg ale i dobre ćwieczenie wpajające dobre praktyki.